src/psbt.c

```diff
@@ -1002,3 +1002,521 @@ static int count_psbt_parts(
+
+            // Read the signature
+            p += varint_from_bytes(p, &value_len);
+            clone_bytes(&partial_sigs->items[partial_sigs->num_items].sig, p, value_len);
+            partial_sigs->items[partial_sigs->num_items].sig_len = value_len;
+
+            partial_sigs->num_items++;
+            p += value_len;
+            break;
+        }
+        case WALLY_PSBT_IN_SIGHASH_TYPE: {
+            if (result->sighash_type > 0) {
+                return WALLY_EINVAL;     // Sighash already provided
+            } else if (key_len != 1) {
+                return WALLY_EINVAL;     // Type is more than one byte
+            }
+            p += varint_from_bytes(p, &value_len);
+            p += uint32_from_le_bytes(p, &result->sighash_type);
+            break;
+        }
+        case WALLY_PSBT_IN_REDEEM_SCRIPT: {
+            if (result->redeem_script_len != 0) {
+                return WALLY_EINVAL;     // Already have a redeem script
+            } else if (key_len != 1) {
+                return WALLY_EINVAL;     // Type is more than one byte
+            }
+            p += varint_from_bytes(p, &value_len);
+            clone_bytes(&result->redeem_script, p, value_len);
+            result->redeem_script_len = value_len;
+
+            p += value_len;
+            break;
+        }
+        case WALLY_PSBT_IN_WITNESS_SCRIPT: {
+            if (result->witness_script_len != 0) {
+                return WALLY_EINVAL;     // Already have a witness script
+            } else if (key_len != 1) {
+                return WALLY_EINVAL;     // Type is more than one byte
+            }
+            p += varint_from_bytes(p, &value_len);
+            clone_bytes(&result->witness_script, p, value_len);
+            result->witness_script_len = value_len;
+
+            p += value_len;
+            break;
+        }
+        case WALLY_PSBT_IN_BIP32_DERIVATION: {
+            struct keypath_map *keypaths = result->keypaths;
+            size_t path_len;
+            if (key_len != 66 && key_len != 34) {
+                return WALLY_EINVAL;     // Size of key is unexpected
+            }
+            // Check for duplicates
+            for (i = 0; i < keypaths->num_items; ++i) {
+                if (memcmp(keypaths->items[i].pubkey, &key[1], key_len - 1) == 0) {
+                    return WALLY_EINVAL;     // Duplicate key
+                }
+            }
+
+            memcpy(keypaths->items[keypaths->num_items].pubkey, &key[1], key_len - 1);
+
+            // Read the path length
+            p += varint_from_bytes(p, &value_len);
+            if (value_len % 4 != 0 || value_len == 0) {
+                return WALLY_EINVAL;     // Invalid length for keypaths
+            }
+            path_len = (value_len / 4) - 1;
+
+            // Read the fingerprint
+            memcpy(keypaths->items[keypaths->num_items].origin.fingerprint, p, 4);
+            p += 4;
```

Comment:
I've added a check after each key and value read that checks that we aren't going to read past the end of the buffer. Instead of an assert, this will return `WALLY_EINVAL`.